### PR TITLE
crimson/common/tri_mutex: make locking/promotion atomic if possible

### PR DESCRIPTION
**Pull request by cyx1231st**
Fixes: https://tracker.ceph.com/issues/65451
Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/osd/object_context.h [hunk 1]
```diff
@@ -121,13 +121,23 @@ class ObjectContext : public ceph::common::intrusive_lru_base<
...
+          lock.unlock();
+        });
       });
-    });
+    } else {
+      return seastar::futurize_invoke(func
+      ).finally([&lock, obc] {
+        lock.unlock();
+      });
+    }
```
Comment (tchaikov) [inline src/crimson/osd/object_context.h:140]: the commit message is missing the problem description,  the root cause analysis, and the summary of this change.  could you please elaborate a little bit on the root cause of

> tri_mutex::promote_from_read(): Assertion `readers == 1' failed.

and how this change addresses this problem?

File: src/crimson/common/tri_mutex.h [hunk 2]
```diff
@@ -51,7 +53,8 @@ class excl_lock_from_excl {
 /// independent of the conventional reader/writer lock usage. Here, what we
 /// mean is that we can pipeline reads, and we can pipeline writes, but we
 /// cannot allow a read while writes are in progress or a write while reads are
-/// in progress. Any rmw operation is therefore exclusive.
+/// in progress. Any rmw operation is therefore exclusive. Note that multiple
+/// writers are allowed by tri_mutex.
```
Comment (tchaikov) [inline src/crimson/common/tri_mutex.h:57]: please put this change in another commit. if we want to emphasis that multiple writes are allowed by `tri_mutex`, probably we could put this way:

> Unlike reader/write lock, tri_mutex does not enforce the exclusive access of write operations, on the contrary, multiple write operations are allowed to hold the same tri_mutex at the same time. Here, what we mean is that we can pipeline reads, and we can pipeline writes, but we cannot allow a read while writes are in progress or a write while reads are in progress.

File: src/crimson/osd/object_context.h (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/osd/object_context.h:140]: My understanding see https://github.com/ceph/ceph/pull/55488#issuecomment-2009065584 that caused `writers == 1` failure.

I think the issue is similar that `func` (to promote and load-obc) isn't called immediately upon the 1st locked. The returned future allows the 2nd reader enter and lock (causing readers > 1) even the future is available, before `func` is called.

Making lock() to return nullopt instead of a ready future in this change makes the 1st locking and calling `func` atomic.

Note that we have fixed several issues related to out-of-order and ready future.

Comment (cyx1231st) [reply src/crimson/osd/object_context.h]: Added more information in the commit message.

File: src/crimson/common/tri_mutex.h (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/common/tri_mutex.h:57]: Splitted.

**Comment by cyx1231st** [comment]
jenkins test api

**Comment by cyx1231st** [comment]
jenkins test api

File: src/crimson/osd/object_context.h (hunk 1 above)
Comment (athanatos) [reply src/crimson/osd/object_context.h:140]: I don't see how this actually helps.  The bulk of the changes here allow for the various lock invocations to *dynamically* return std::nullopt allowing us to take the else branch of the if statement above

```
      return seastar::futurize_invoke(func
      ).finally([&lock, obc] {
        lock.unlock();
      });
```

That only solves the atomicity problem it if actually happens to return std::nullopt, though.  For this to be an actual solution, we'd have to know statically that it really will.  In that case, why allow the function to return a non-ready future at all?

I think this is essentially symmetric to the other similar PR that attempted to return a std::optional<future<>> -- the right solution is to identify why we are confident that a particular pathway doesn't need to block and directly call a variant that will never block.

Comment (cyx1231st) [reply src/crimson/osd/object_context.h:140]: I see, if my understanding is correct above, we should be able to convert read/write promotion to a non-block version, will try.

BTW, the test with the current change looks passing, see https://github.com/ceph/ceph/pull/56912#issuecomment-2060278568

Comment (cyx1231st) [reply src/crimson/osd/object_context.h]: The original promotion is already synchronous, so spcialize `_with_promotion()` to make sure the function is called atomically.

Proceed to tests.

**Review by athanatos — APPROVED** [review]
Yeah, that's simpler.

**Comment by cyx1231st** [comment]
jenkins retest this please

File: src/crimson/common/tri_mutex.h [hunk 3]
```diff
@@ -27,22 +27,15 @@ class excl_lock {
 // promote from read to excl
 class excl_lock_from_read {
 public:
-  seastar::future<> lock();
-  void unlock();
+  void promote();
+  void demote();
```
Comment (tchaikov) [inline src/crimson/common/tri_mutex.h:31]: this change is confusing. the naming of `lock` and `unlock` follows the convention of the lock concept in standard library. and it reflect the semantics of the "lock" in `excl_lock_from_read`. i would vote to keep the name, if i could.

File: src/crimson/common/tri_mutex.h (hunk 2 above)
Comment (tchaikov) [reply src/crimson/common/tri_mutex.h]: could you reconsider the suggestion above?

File: src/crimson/common/tri_mutex.h (hunk 3 above)
Comment (cyx1231st) [reply src/crimson/common/tri_mutex.h:31]: I don't have strong preference here. I'm good to change back to the names lock/unlock.

Comment (cyx1231st) [reply src/crimson/common/tri_mutex.h:31]: But I still want to differentiate `lock() -> future<>` vs `lock() -> void`. How about rename to `lock_sync()` or `lock_excl()` ?

File: src/crimson/common/tri_mutex.h (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/common/tri_mutex.h]: Yes, updated.

File: src/crimson/common/tri_mutex.h (hunk 3 above)
Comment (cyx1231st) [reply src/crimson/common/tri_mutex.h]: Renamed back to lock() and protect with concepts instead.

File: src/crimson/common/tri_mutex.cc [hunk 4]
```diff
@@ -3,6 +3,8 @@
 
 #include "tri_mutex.h"
 
+#include <seastar/util/later.hh>
```
Comment (tchaikov) [inline src/crimson/common/tri_mutex.cc:6]: why are we include this header?

File: src/crimson/common/tri_mutex.cc [hunk 5]
```diff
@@ -72,7 +63,7 @@ tri_mutex::~tri_mutex()
 seastar::future<> tri_mutex::lock_for_read()
 {
   if (try_lock_for_read()) {
-    return seastar::make_ready_future<>();
+    return seastar::now();
```
Comment (tchaikov) [inline src/crimson/common/tri_mutex.cc:66]: this change does not belong to this commit.

File: src/crimson/osd/object_context.h [hunk 6]
```diff
@@ -121,7 +121,7 @@ class ObjectContext : public ceph::common::intrusive_lru_base<
   bool recovery_read_marker = false;
 
   template <typename Lock, typename Func>
-  auto _with_lock(Lock&& lock, Func&& func) {
+  auto _with_lock(Lock &lock, Func &&func) {
```
Comment (tchaikov) [inline src/crimson/osd/object_context.h:124]: i understand that we've been putting `&` near the variable. but `&` is part of the type, see https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines for more examples, so..

File: src/crimson/osd/object_context.h [hunk 7]
```diff
@@ -130,6 +130,15 @@ class ObjectContext : public ceph::common::intrusive_lru_base<
...
   }
 
+  template <TriMutexSyncLock Lock, typename Func>
+  auto _with_promoted_lock(Lock &lock, Func &&func) {
+    Ref obc = this;
+    lock.lock();
+    return seastar::futurize_invoke(func).finally([&lock, obc] {
+      lock.unlock();
+    });
+  }
```
Comment (tchaikov) [inline src/crimson/osd/object_context.h:140]: could put something like
```c++
template <typename Mutex, typename Func>
auto _with_lock(Mutex& mutex, Func&& func) {
    Ref obc{this};
    const std::lock_guard<Mutex> lock{mutex};
    return seastar::futurize_invoke(std::forward(func));
}
```

IMHO, no need to use the concept for enforcing the type constraints. because `future<>` cannot be discarded with `[[nodiscard]]` attribute. i think SFINAE should work here, so no need to use a new name, unless we need to make it more visible. but for what reason?

File: src/crimson/common/tri_mutex.h [hunk 8]
```diff
@@ -3,9 +3,15 @@
...
 #pragma once
 
+#include <concepts>
+
 #include <seastar/core/future.hh>
 #include <seastar/core/circular_buffer.hh>
 
+template <typename Lock>
+concept TriMutexSyncLock = std::is_void_v<
+  std::invoke_result_t<decltype(&Lock::lock), Lock> >;
```
Comment (tchaikov) [inline src/crimson/common/tri_mutex.h:13]: if we have to add a concept for [BasicLockable](http://www.cplusplus.com/reference/concept/BasicLockable/). might want to put something like:
```c++
template<typename Mutex>
concept basic_lockable = requires(Mutex m) {
    { m.lock()   } -> std::same_as<void>;
    { m.unlock() } -> std::same_as<void>;
};
```

**Comment by tchaikov** [comment]
could s/if available/if possible/ or s/if available/when appropriate/ in the title of this PR.

File: src/crimson/common/tri_mutex.cc (hunk 4 above)
Comment (cyx1231st) [reply src/crimson/common/tri_mutex.cc:6]: It's for seastar::now().

File: src/crimson/osd/object_context.h (hunk 7 above)
Comment (cyx1231st) [reply src/crimson/osd/object_context.h:140]: Seems not working directly:
```
/root/ceph/src/crimson/osd/object_context.h:138:15: error: ‘co_await’ cannot be used in a function with a deduced return type
  138 |     co_return co_await seastar::futurize_invoke(func);
      |               ^~~~~~~~
/root/ceph/src/crimson/osd/object_context.h:135:8: error: unable to find the promise type for this coroutine
  135 |   auto _with_promoted_lock(Mutex& mutex, Func&& func) {
```

Will fallback to the original version.

File: src/crimson/common/tri_mutex.h (hunk 8 above)
Comment (cyx1231st) [reply src/crimson/common/tri_mutex.h]: Discarded the concept.

File: src/crimson/common/tri_mutex.cc (hunk 5 above)
Comment (cyx1231st) [reply src/crimson/common/tri_mutex.cc:66]: splitted

**Comment by cyx1231st** [comment]
See https://pulpito.ceph.com/yingxin-2024-04-23_01:23:55-crimson-rados-ci-yingxin-crimson-fix-tri-mutex-4-distro-default-smithi/

All look good @Matan-B @athanatos

**Comment by cyx1231st** [comment]
jenkins test api

**Review by tchaikov — APPROVED** [review]
lgtm

File: src/crimson/osd/object_context.h (hunk 7 above)
Comment (tchaikov) [reply src/crimson/osd/object_context.h:140]: thanks. will try to come up with a followup.
